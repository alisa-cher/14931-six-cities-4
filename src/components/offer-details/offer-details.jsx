import React from "react";
import PropTypes from "prop-types";
import CommentsList from "../comments-list/comments-list.jsx";
import OffersList from "../offers-list/offers-list.jsx";
import ReviewForm from "../review-form/review-form.jsx";
import Map from "../map/map.jsx";
import {sliceAnArray} from "../../helpers.js";
import {detailedOfferShape, offerShape} from "../../types.js";
import withActiveItem from "../../hocs/with-active-item/with-active-item.jsx";
import withToggleItem from "../../hocs/with-toggle-item/with-toggle-item.jsx";

const ReviewFormWrapped = withToggleItem(withActiveItem(ReviewForm));

const OfferDetails = (props) => {
  const {
    isSendReviewError,
    isUserLoggedIn,
    onSubmit,
    offer,
    activeItem,
    nearbyOffers,
    comments,
    onCardTitleClick,
    setActiveItem,
    resetActiveItem,
    cityCoords,
    cityZoom
  } = props;

  const {title, description, rating, isPremium, photos, price, type, goods, host, bedrooms, maxAdults, id} = offer;
  const {avatar, isPro, name} = host;
  const commentsLength = comments.length;
  const NEARBY_OFFERS_MAX_AMOUNT = 3;
  const COMMENTS_MAX_AMOUNT = 10;
  const offersToRender = sliceAnArray(nearbyOffers, NEARBY_OFFERS_MAX_AMOUNT);
  const commentsToRender = sliceAnArray(comments, COMMENTS_MAX_AMOUNT);
  const sortedComments = commentsToRender.slice().sort((a, b) => new Date(b.date) - new Date(a.date));

  return (
    <div className="page">
      <header className="header">
        <div className="container">
          <div className="header__wrapper">
            <div className="header__left">
              <a className="header__logo-link" href="main.html">
                <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width="81" height="41"/>
              </a>
            </div>
            <nav className="header__nav">
              <ul className="header__nav-list">
                <li className="header__nav-item user">
                  <a className="header__nav-link header__nav-link--profile" href="#">
                    <div className="header__avatar-wrapper user__avatar-wrapper">
                    </div>
                    <span className="header__user-name user__name">Oliver.conner@gmail.com</span>
                  </a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </header>
      <main className="page__main page__main--property">
        <section className="property">
          <div className="property__gallery-container container">
            <div className="property__gallery">
              {photos.map((photo) =>
                <div className="property__image-wrapper" key={photo}>
                  <img className="property__image" src={photo} alt="Photo studio"/>
                </div>
              )}
            </div>
          </div>
          <div className="property__container container">
            <div className="property__wrapper">
              {isPremium && <div className="property__mark">
                <span>Premium</span>
              </div>}
              <div className="property__name-wrapper">
                <h1 className="property__name">
                  {title}
                </h1>
                <button className="property__bookmark-button button" type="button">
                  <svg className="property__bookmark-icon" width="31" height="33">
                    <use xlinkHref="#icon-bookmark"/>
                  </svg>
                  <span className="visually-hidden">To bookmarks</span>
                </button>
              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={{width: `80%`}}/>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">{rating}</span>
              </div>
              <ul className="property__features">
                <li className="property__feature property__feature--entire">
                  {type}
                </li>
                <li className="property__feature property__feature--bedrooms">
                  {bedrooms > 1 ? bedrooms + ` Bedrooms` : bedrooms + ` Bedroom`}
                </li>
                <li className="property__feature property__feature--adults">
                  Max {maxAdults > 1 ? maxAdults + ` adults` : maxAdults + ` adult`}
                </li>
              </ul>
              <div className="property__price">
                <b className="property__price-value">&euro;{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <div className="property__inside">
                <h2 className="property__inside-title">What&apos;s inside</h2>
                <ul className="property__inside-list">
                  {goods.map((item) =>
                    <li key={item} className="property__inside-item">
                      {item}
                    </li>
                  )}
                </ul>
              </div>
              <div className="property__host">
                <h2 className="property__host-title">Meet the host</h2>
                <div className="property__host-user user">
                  <div className={`property__avatar-wrapper user__avatar-wrapper ${isPro ? `` : `property__avatar-wrapper--pro`}`}>
                    <img className="property__avatar user__avatar" src={avatar} width="74" height="74"
                      alt="Host avatar"/>
                  </div>
                  <span className="property__user-name">
                    {name}
                  </span>
                </div>
                <div className="property__description">
                  <p className="property__text">
                    {description}
                  </p>
                </div>
              </div>
              <section className="property__reviews reviews">
                <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{commentsLength}</span></h2>
                <CommentsList comments={sortedComments}/>
                {isUserLoggedIn &&
                  <ReviewFormWrapped
                    onSubmit={onSubmit}
                    offerId={id}
                    isError={isSendReviewError}
                  />}
              </section>
            </div>
          </div>
          <section className="property__map map">
            <Map
              offers={offersToRender}
              detailedOffer={offer}
              activeCard={activeItem}
              cityCoords={cityCoords}
              cityZoom={cityZoom}
            />
          </section>
        </section>
        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">Other places in the neighbourhood</h2>
            <OffersList
              offers={offersToRender}
              onCardTitleClick={onCardTitleClick}
              onCardHover={setActiveItem}
              onCardMouseLeave={resetActiveItem}
              classNamePrefix={`near-places`}
            />
          </section>
        </div>
      </main>
    </div>
  );
};

OfferDetails.propTypes = {
  isUserLoggedIn: PropTypes.bool,
  isSendReviewError: PropTypes.bool,
  offer: PropTypes.shape(detailedOfferShape).isRequired,
  activeItem: PropTypes.object,
  nearbyOffers: PropTypes.arrayOf(PropTypes.shape(offerShape)).isRequired,
  comments: PropTypes.array.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCardTitleClick: PropTypes.func.isRequired,
  cityCoords: PropTypes.arrayOf(PropTypes.number).isRequired,
  cityZoom: PropTypes.number.isRequired,
  setActiveItem: PropTypes.func.isRequired,
  resetActiveItem: PropTypes.func.isRequired
};

export default OfferDetails;
