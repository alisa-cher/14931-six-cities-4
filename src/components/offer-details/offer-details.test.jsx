import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from "react-router-dom";
import {OfferDetails} from '../offer-details/offer-details.jsx';

jest.mock(`../map/map`, () => jest.fn().mockReturnValue(null));

const offer = {
  id: 0,
  bedrooms: 3,
  description: `An independent House`,
  maxAdults: 4,
  type: `appartement`,
  price: 140,
  title: `just amazing1`,
  isPremium: true,
  isFavorite: true,
  rating: 4.3,
  photos: [`img/room.jpg`, `img/studio-01.jpg`],
  goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
  host: {
    avatar: `img/avatar.svg`,
    isPro: true,
    name: `Angelina`
  },
  previewPhoto: `img/apartment-01.jpg`,
};

const comments = [{
  id: 4,
  user: {
    name: `Emely`,
    avatarUrl: `https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/8.jpg`},
  rating: 4,
  comment: `Beautiful space, fantastic location and atmosphere, really a wonderful place to spend a few days. Will be back.`,
  date: `2020-06-17T10:54:44.806Z`
}];

it(`Property page renders correctly`, () => {
  const tree = renderer
    .create(
        <BrowserRouter>
          <OfferDetails
            isUserLoggedIn={true}
            nearbyOffers={[offer]}
            comments={comments}
            onCardTitleClick={() => {}}
            onCardHover={() => {}}
            onCardMouseLeave={() => {}}
            cityCoords={[0, 2]}
            cityZoom={1}
            onSubmit={() => {}}
            onItemReset={() => {}}
            onItemSet={() => {}}
            hotelId={0}
            offers={[offer]}
            onFavoriteButtonClick={() => {}}
            onGetComments={() => {}}
            onGetNearbyOffers={() => {}}
          />
        </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
