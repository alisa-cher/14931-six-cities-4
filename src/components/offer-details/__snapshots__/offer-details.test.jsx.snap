// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property page renders correctly 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="../img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                >
                  <img
                    alt="user avatar"
                    height="41"
                    style={
                      Object {
                        "borderRadius": "10px",
                      }
                    }
                    width="81"
                  />
                </div>
                <span
                  className="header__user-name user__name"
                />
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/room.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/studio-01.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__mark"
          >
            <span>
              Premium
            </span>
          </div>
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              just amazing1
            </h1>
            <button
              className="button property__bookmark-button property__bookmark-button--active"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="button property__bookmark-icon"
                height={33}
                width={32}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__rating property"
            >
              <div
                className="property__stars rating__stars"
              >
                <span
                  style={
                    Object {
                      "width": "80%",
                    }
                  }
                />
                <span
                  className="visually-hidden"
                >
                  Rating
                </span>
              </div>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              4.3
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              Appartement
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              3 Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              4 adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              140
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Heating
              </li>
              <li
                className="property__inside-item"
              >
                Kitchen
              </li>
              <li
                className="property__inside-item"
              >
                Cable TV
              </li>
              <li
                className="property__inside-item"
              >
                Washing machine
              </li>
              <li
                className="property__inside-item"
              >
                Coffee machine
              </li>
              <li
                className="property__inside-item"
              >
                Dishwasher
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper user__avatar-wrapper "
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="../img/avatar.svg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                An independent House
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              >
                1
              </span>
            </h2>
            <ul
              className="reviews__list"
            >
              <li
                className="reviews__item"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      src="https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/8.jpg"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Emely
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating reviews"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    Beautiful space, fantastic location and atmosphere, really a wonderful place to spend a few days. Will be back.
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2020-6-17"
                  >
                    June 17, 2020 
                  </time>
                </div>
              </li>
            </ul>
            <form
              action="#"
              className="reviews__form form"
              method="post"
              onSubmit={[Function]}
            >
              <label
                className="reviews__label form__label"
                htmlFor="review"
              >
                Your review
              </label>
              <div
                className="reviews__rating-form form__rating"
              >
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  id="5-stars"
                  name="rating"
                  onChange={[Function]}
                  type="radio"
                  value={5}
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="5-stars"
                  title="perfect"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  id="4-stars"
                  name="rating"
                  onChange={[Function]}
                  type="radio"
                  value={4}
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="4-stars"
                  title="good"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  id="3-stars"
                  name="rating"
                  onChange={[Function]}
                  type="radio"
                  value={3}
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="3-stars"
                  title="not bad"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  id="2-stars"
                  name="rating"
                  onChange={[Function]}
                  type="radio"
                  value={2}
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="2-stars"
                  title="badly"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  checked={false}
                  className="form__rating-input visually-hidden"
                  id="1-stars"
                  name="rating"
                  onChange={[Function]}
                  type="radio"
                  value={1}
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="1-stars"
                  title="terribly"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
              </div>
              <textarea
                className="reviews__textarea form__textarea"
                id="review"
                maxLength="500"
                minLength="30"
                name="review"
                onChange={[Function]}
                placeholder="Tell how was your stay, what you like and what can be improved"
              />
              <div
                className="reviews__button-wrapper"
              >
                <p
                  className="reviews__help"
                >
                  To submit review please make sure to set 
                  <span
                    className="reviews__star"
                  >
                    rating
                  </span>
                   and describe your stay with at least 
                  <b
                    className="reviews__text-amount"
                  >
                    50 characters
                  </b>
                  .
                </p>
                <button
                  className="reviews__submit form__submit button"
                  disabled={true}
                  type="submit"
                >
                  Submit
                </button>
              </div>
            </form>
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      />
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="places__list near-places__list"
        >
          <article
            className="place-card near-places__card"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="place-card__mark"
            >
              <span>
                Premium
              </span>
            </div>
            <div
              className="near-places__image-wrapper place-card__image-wrapper"
            >
              <a
                href="/offer/0"
                onClick={[Function]}
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height="200"
                  src="img/apartment-01.jpg"
                  width="260"
                />
              </a>
            </div>
            <div
              className="place-card__info"
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    140
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="button place-card__bookmark-button place-card__bookmark-button--active"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="button place-card__bookmark-icon"
                    height={19}
                    width={18}
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    To bookmarks
                  </span>
                </button>
              </div>
              <div
                className="place-card__rating rating"
              >
                <div
                  className="place-card__rating place-card"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "80%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
              </div>
              <h2
                className="place-card__name"
                onClick={[Function]}
              >
                <a
                  href="/offer/0"
                  onClick={[Function]}
                >
                  just amazing1
                </a>
              </h2>
              <p
                className="place-card__type"
              >
                Appartement
              </p>
            </div>
          </article>
        </div>
      </section>
    </div>
  </main>
</div>
`;
